tutorial: nuttx: update micro-ROS over 6lowpan (#213)
micro-ROS/micro-ROS.github.io: _docs/tutorials/advanced/nuttx/6lowpan/index.md

# What is 6LoWPAN?

6LoWPAN is an acronym of IPv6 over Low-Power Wireless Personal Area Networks.
This communication protocol allows wireless communication over IEEE 802.15.4 based networks using IPv6. 
Some of the main advantages are:

- Easy to route from radio devices to the Internet, thanks to the usage of the IP packets.
- Easy to use on UDP and TCP server/clients.
- A protocol designed for low power and constrained devices. Perfect for Micro-ROS remote sensors.

# Needed devices

At this moment this communication protocol is only available for NuttX RTOS. 
Below the list of what is needed to continue:

- Olimex-STM32-E407 board. [Link](https://www.olimex.com/Products/ARM/ST/STM32-E407/open-source-hardware)
- PmodRF2 Radio. [Link](https://store.digilentinc.com/pmod-rf2-ieee-802-15-rf-transceiver/)
- A Linux Ubuntu 18.04 machine with docker installed.
- A 802.15.4 Radio USB dongle [Link](http://shop.sysmocom.de/products/atusb) plugged into the Linux machine.

# Configure the Olimex board

The configuration of the board is divided into two parts: the set-up of the hardware and the set-up of the software.

## Set-up the hardware

> First the PmodRF2 need to be connected to the Olimex board as desribed below:

Reviewer: ```suggestion
First the PmodRF2 needs to be connected to the Olimex board as desribed below:
```

- `Board D13` -> `MRF24J40 SCLK`
- `Board D12` -> `MRF24J40 MISO`
- `Board D11` -> `MRF24J40 MOSI`
- `Board D10` -> `MRF24J40 CS`,
- `Board D8` -> `MRF24J40 INT`.

The ``D`` letter refers to the Arduino header pins of the board denoted ``CON4``.
To ease the set-up process, The radio's pinout can be found at the following link: [PmodRF2 pinout](https://reference.digilentinc.com/reference/pmod/pmodrf2/start)

The last step is to connect a mini-USB cable to the OTG2 USB port (The USB port next to the Ethernet port).

## Set-up the software

> The micro-ROS build system will be use in order to create and flash the

Reviewer: ```suggestion
The micro-ROS build system will be used in order to create and flash the
```

firmware.

> Instructions are available at the following link: [Micro-ROS build system](https://github.com/micro-ROS/micro-ros-build/blob/dashing/micro_ros_setup/README.md).

Reviewer: ```suggestion
Instructions are available at the following link: [Micro-ROS build system](https://github.com/micro-ROS/micro_ros_setup/blob/dashing/micro_ros_setup/README.md).
```

Reviewer 2: Interesting, both links works for some reason.

For this particular guide, it is necessary to use the branch ``dashing`` and the configuration profile ``uros_6lowpan``.

### Manual patches

> Due to the build system being a work in progress, some of files have to be

Reviewer: ```suggestion
Due to the build system being a work in progress, some of the files have to be
```

modified.

- Open ``uros_ws/firmware/mcu_ws/eProsima/Micro-XRCE-DDS-Client/client.config`` and modify the value of ``CONFIG_UDP_TRANSPORT_MTU`` from ``512`` to ``450``.

This is an appropiate MTU for 6LoWPAN communications on NuttX.

> Also the micro-XRCE-DDS needs to be to be change to the version 1.2.5 in the docker:

Reviewer: ```suggestion
Also the Micro-XRCE-DDS library needs to be changed to the version 1.2.5 in the docker:
```

```bash
cd /uros_ws/firmware/mcu_ws/eProsima/Micro-XRCE-DDS-Client
git checkout v1.2.5
```

> Once the modifications done, the binary can be compiled [as explained here](https://github.com/micro-ROS/micro_ros_setup/blob/dashing/micro_ros_setup/README.md)

Reviewer: ```suggestion
Once the above modifications are done, the binary can be compiled [as explained here](https://github.com/micro-ROS/micro_ros_setup/blob/dashing/micro_ros_setup/README.md)
```

# Configure the Linux machine

> The Linux machine is where is going to run the agent.

Reviewer: ```suggestion
The Linux machine is where the agent is going to be run.
```

> In order to setup the wireless, some settings have to be set as follow:

Reviewer: ```suggestion
In order to setup the wireless transport, some settings have to be set as follows:
```

```bash
#Setting up the 6lowpan network
sudo apt update && sudo apt install iwpan # install the iwpan tool
sudo ip link set wpan0 down
sudo ip link set lowpan0 down
sudo iwpan dev wpan0 set pan_id 0xabcd
sudo iwpan phy phy0 set channel 0 26 # phy can be phy0 or any other phyN where N [0,1,2,3,4,5,6,7,8,9,10,...]
sudo ip link add link wpan0 name lowpan0 type lowpan
sudo ip link set wpan0 up
sudo ip link set lowpan0 up
sudo ip link set docker0 down # Necessary to not reroute everything over the docker and use the lowpan0
sudo ip neigh flush all
sudo ip neigh add to {**IPV6 of the microROS node**} dev lowpan0 lladdr {**HW 802.15.4 of the micro-controller**} # repeat it for all the device
```

It appears that depending on the PC configuration, the packets will not be able to be routed correctly from the docker to the lowpan0/wpan0 and vice-versa.
In order to forward the packet to the correct interface the following commands need to be entered: 

```bash
#Setting up the 6lowpan routing
sudo ip -6 route add {**IPV6 of the microROS node**} dev lowpan0 proto kernel metric 50 pref medium
```

> Once hte setup done, the docker may be run:

Reviewer: ```suggestion
Once the setup is done, the docker may be run:
```

```bash
docker run -it --network host microros/micro-ros-agent:dashing udp6 --port 9999  -v6

[1599566641.351706] info     | UDPv6AgentLinux.cpp | init                     | running...             | port: 9999
[1599566641.351889] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 6

```

# How to use it?

> - Olimex-STM32-E407 board must be switched on and a connection to the NSH

Reviewer: ```suggestion
- The Olimex-STM32-E407 board must be switched on and a connection to the NSH
```

Builtin Apps:
  ping6         i8sak         uros_6lowpan 

```

>  - Then the publisher can executed as follow. Note that the ipv6 address (fe80::bc81:c3b9:5c14:1ab) is the

Reviewer: ```suggestion
 - Then the publisher can be executed as follow. Note that the IPv6 address (fe80::bc81:c3b9:5c14:1ab) is the
```

>    ipv6 address of the Linux lowpan0 interface.

Reviewer: ```suggestion
   IPv6 address of the Linux lowpan0 interface.
```
